packages/solo/src/chain-cosmos-sdk.js

```diff
@@ -381,137 +342,220 @@ ${chainID} chain does not yet know of address ${myAddr}${adviseEgress(myAddr)}
   // Begin the block notifier cycle.
   const blockNotifier = getBlockNotifier();
 
-  /**
-   * This function is entered at most the same number of times
-   * as the blockNotifier announces a new block.
-   *
-   * It then gets the mailbox.  There are no optimisations.
-   *
-   * @param {number=} lastBlockUpdate
-   */
-  const recurseEachNewBlock = (lastBlockUpdate = undefined) => {
-    blockNotifier
-      .getUpdateSince(lastBlockUpdate)
-      .then(({ updateCount, value }) => {
-        assert(value, X`${GCI} unexpectedly finished!`);
-        log.debug(`new block on ${GCI}, fetching mailbox`);
-        getMailbox()
-          .then(ret => {
-            if (!ret) {
-              // The getMailbox was cancelled.
-              return;
-            }
-
-            const { outbox, ack } = ret;
-            // console.debug('have outbox', outbox, ack);
-            inbound(GCI, outbox, ack);
-          })
-          .catch(e => log.error(`Failed to fetch ${GCI} mailbox:`, e));
-        recurseEachNewBlock(updateCount);
-      });
-  };
+  const { notifier: sendNotifier, updater: sendUpdater } = makeNotifierKit();
 
-  // Begin the block consumer.
-  recurseEachNewBlock();
+  // An array of [seqnum, message] ordered by unique seqnum.
+  let messagePool = [];
+
+  // Atomically add to the message pool, ensuring the pool is sorted by unique
+  // sequence number.
+  const addToMessagePool = newMessages => {
+    // Add the new messages.
+    const bigPool = messagePool.concat(newMessages);
+
+    // Sort the big pool by sequence number.
+    const sortedPool = bigPool.sort((a, b) => a[0] - b[0]);
+
+    // Only keep messages that have a unique sequence number.
+    const uniquePool = sortedPool.filter(
+      (a, i) =>
+        // Index 0 is always kept.
+        i === 0 ||
+        // Others are only kept if not the same as the prior sequence.
+        a[0] !== sortedPool[i - 1][0],
+    );
+
+    // Replace the old message pool.
+    messagePool = uniquePool;
+  };
 
   let totalDeliveries = 0;
-  async function deliver(newMessages, acknum) {
+  let highestAck = -1;
```

Comment:
It's surprising that `highestAck` is a Number, but `sequenceNumber` is a BigInt. I'm personally not very consistent about using one vs the other, so I suspect this is my fault, but it might be worth knowing why they differ.

Reply:
That's my typing laziness of `highestAck`.  `0n < 1` and `0 < 1n`.  I'll look to find out what the ack actually is.

Reply:
I'd prefer to use `BigInt` for things that really are ints, so that's why I did for sequenceNumber.  Looks like ack is just a `number`, so I'll leave `highestAck` to be that.